crates/red_knot_ide/src/inlay_hints.rs

```diff
@@ -0,0 +1,193 @@
+
+    let hints = visitor.hints().clone();
+
+    hints
+}
+
+struct InlayHintVisitor<'db> {
+    model: SemanticModel<'db>,
+    file: File,
+    hints: Vec<RangedValue<InlayHintContent<'db>>>,
+}
+
+impl<'db> InlayHintVisitor<'db> {
+    fn new(db: &'db dyn Db, file: File) -> Self {
+        Self {
+            model: SemanticModel::new(db.upcast(), file),
+            file,
+            hints: Vec::new(),
+        }
+    }
+
+    fn hints(&self) -> &Vec<RangedValue<InlayHintContent<'db>>> {
+        &self.hints
+    }
+
+    fn add_hint(&mut self, range: TextRange, ty: Type<'db>) {
+        self.hints.push(RangedValue {
+            range: FileRange::new(self.file, range),
+            value: InlayHintContent::AssignStatement(ty),
+        });
+    }
+}
+
+impl SourceOrderVisitor<'_> for InlayHintVisitor<'_> {
+    fn visit_stmt(&mut self, stmt: &Stmt) {
+        match stmt {
+            Stmt::Assign(assign) => {
+                let ty = assign.value.inferred_type(&self.model);
+                for target in &assign.targets {
+                    match target {
+                        Expr::Tuple(tuple) => {
+                            for element in &tuple.elts {
+                                let element_ty = element.inferred_type(&self.model);
+                                self.add_hint(element.range(), element_ty);
+                            }
+                        }
+                        _ => {
+                            self.add_hint(target.range(), ty);
+                        }
+                    }
+                }
+                return;
+            }
+            // TODO
+            Stmt::FunctionDef(_) => {}
+            Stmt::For(_) => {}
+            Stmt::Expr(_) => {}
+            _ => {}
+        }
+
+        source_order::walk_stmt(self, stmt);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use red_knot_python_semantic::types::StringLiteralType;
+    use ruff_db::files::{system_path_to_file, File};
+    use ruff_db::system::DbWithWritableSystem as _;
+    use ruff_text_size::TextSize;
+
+    use crate::db::tests::TestDb;
+
+    struct TestCase {
+        db: TestDb,
+        file: File,
+    }
+
+    fn test_case(content: impl AsRef<str>) -> TestCase {
+        let mut db = TestDb::new();
+        db.write_file("test.py", content).unwrap();
+
+        let file = system_path_to_file(&db, "test.py").unwrap();
+
+        TestCase { db, file }
+    }
+
+    #[test]
+    fn test_assign_statement() {
+        let test_case = test_case("x = 1");
+        let hints = get_inlay_hints(&test_case.db, test_case.file);
+        assert_eq!(hints.len(), 1);
+        assert_eq!(
+            hints[0].value,
+            InlayHintContent::AssignStatement(Type::IntLiteral(1))
+        );
+        assert_eq!(
+            hints[0].range,
+            FileRange::new(
+                test_case.file,
+                TextRange::new(TextSize::from(0), TextSize::from(1))
+            )
+        );
```

Comment:
Could we use `Diagnostic` rendering with inline `insta` snapshots here similar to the goto type definition and on-hover tests? Tests with text ranges are always very hard to read and the diagnostic rendering can help because it visualizes the ranges (much easier to spot errors)